include/seqan3/io/structure/structure_file_in.hpp

```diff
@@ -0,0 +1,1217 @@
+ * \tparam stream_type        The type of the stream, must satisfy seqan3::istream_concept.
+ * \details
+ *
+ * ### Introduction
+ *
+ * Structured sequence files contain intra-molecular interactions of RNA or protein. Usually, but not necessarily, they
+ * contain the nucleotide or amino acid sequences and descriptions as well. Interactions can be represented
+ * either as fixed _secondary structure_, where every character is assigned at most one interaction partner
+ * (structure of minimum free energy), or an _annotated sequence_, where every character is assigned a set
+ * of interaction partners with specific base pair probabilities.
+ *
+ * The structured sequence file abstraction supports reading ten different fields:
+ *
+ *   1. seqan3::field::SEQ (sequence)
+ *   2. seqan3::field::ID (identifier)
+ *   3. seqan3::field::BPP (annotated sequence)
+ *   4. seqan3::field::STRUCTURE (secondary structure)
+ *   5. seqan3::field::STRUCTURED_SEQ (sequence and structure in one range)
+ *   6. seqan3::field::ENERGY (minimum free energy)
+ *   7. seqan3::field::REACT (reactivity)
+ *   8. seqan3::field::REACT_ERR (reactivity error)
+ *   9. seqan3::field::COMMENT (free text)
+ *   10. seqan3::field::OFFSET (index of first sequence character)
+ *
+ * The first three fields are retrieved by default (and in that order). The seqan3::field::STRUCTURED_SEQ may be
+ * selected to have sequence and structure directly stored in a more memory-efficient combined container.
+ * If you select this field you must not select seqan3::field::SEQ or seqan3::field::STRUCTURE.
+ *
+ * ### Construction and specialisation
+ *
+ * This class comes with two constructors, one for construction from a file name and one for construction from
+ * an existing stream and a known format. The first one automatically picks the format based on the extension
+ * of the file name. The second can be used if you have a non-file stream, like std::cin or std::istringstream,
+ * that you want to read from and/or if you cannot use file-extension based detection, but know that your input
+ * file has a certain format.
+ *
+ * In most cases the template parameters are deduced completely automatically:
+ * ```cpp
+ * structure_file_in sf{"/tmp/my.dbn"}; // Dot bracket with RNA sequences assumed, regular std::ifstream taken as stream
+ * ```
+ *
+ * Reading from an std::istringstream:
+ * ```cpp
+ * std::string input
+ * {
+ *     "> S.cerevisiae_tRNA-PHE M10740/1-73\n"
+ *     "GCGGAUUUAGCUCAGUUGGGAGAGCGCCAGACUGAAGAUUUGGAGGUCCUGUGUUCGAUCCACAGAAUUCGCA\n"
+ *     "(((((((..((((........)))).((((.........)))).....(((((.......)))))))))))). (-17.50)\n"
+ *     "> example\n"
+ *     "UUGGAGUACACAACCUGUACACUCUUUC\n"
+ *     "..(((((..(((...)))..)))))... (-3.71)\n"
+ * };
+ *
+ * std::istringstream iss(input);
+ *
+ * structure_file_in fin{std::move(iss), structure_file_format_dot_bracket{}};
+ * //              ^ no need to specify the template arguments
+ * ```
```

Comment:
BTW, you can now use code-snippets from tests, as well, I don't know if that would be helpful for you. See https://github.com/seqan/seqan3/blob/master/include/seqan3/range/detail/inherited_iterator_base.hpp with an example of how it works.